src/app/modules/submission/encrypt-submission/encrypt-submission.middleware.ts

```diff
@@ -87,3 +116,264 @@ export const validateStorageSubmissionParams = celebrate({
+      message: 'Error encrypting attachment',
+      meta: {
+        action: 'encryptAttachment',
+        label,
+        error,
+      },
+    })
+    throw error
+  }
+}
+
+const getEncryptedAttachmentsMapFromAttachmentsMap = async (
+  attachmentsMap: Record<string, Buffer>,
+  publicKey: string,
+): Promise<StorageModeAttachmentsMap> => {
+  const attachmentPromises = Object.entries(attachmentsMap).map(
+    ([id, attachment]) => encryptAttachment(attachment, { id, publicKey }),
+  )
+
+  return Promise.all(attachmentPromises).then((encryptedAttachmentsMeta) =>
+    chain(encryptedAttachmentsMeta)
+      .keyBy('id')
+      // Remove id from object.
+      .mapValues((v) => omit(v, 'id'))
+      .value(),
+  )
+}
+
+/**
+ * Encrypt submission content before saving to DB.
+ */
+export const encryptSubmission: StorageSubmissionMiddlewareHandlerType = async (
+  req,
+  res,
+  next,
+) => {
+  const formId = req.params.formId
+  const encryptedFormDef = req.body.formsg.encryptedFormDef
+  const publicKey = encryptedFormDef.publicKey
+
+  const attachmentsMap: Record<string, Buffer> = {}
+
+  const logMeta = {
+    action: 'encryptSubmission',
+    ...createReqMeta(req),
+    formId,
+  }
+
+  // Populate attachment map
+  req.body.responses.filter(isAttachmentResponse).forEach((response) => {
+    const fieldId = response._id
+    attachmentsMap[fieldId] = response.content
+  })
+
+  const encryptedAttachments =
+    await getEncryptedAttachmentsMapFromAttachmentsMap(
+      attachmentsMap,
+      publicKey,
+    )
+
+  const filteredResponses = getFilteredResponses(
+    encryptedFormDef,
+    req.body.responses,
+  )
+
+  if (filteredResponses.isErr()) {
+    logger.warn({
+      message: filteredResponses.error.message,
+      meta: logMeta,
+    })
+    return res.status(StatusCodes.INTERNAL_SERVER_ERROR).json({
+      message: filteredResponses.error.message,
+    })
+  }
+
+  const strippedBodyResponses = req.body.responses.map((response) => {
+    if (isAttachmentResponse(response)) {
+      return {
+        ...response,
+        filename: undefined,
+        content: undefined, //Strip out attachment content
+      }
+    } else {
+      return response
+    }
+  })
+
+  const encryptedContent = formsgSdk.crypto.encrypt(
+    strippedBodyResponses,
+    publicKey,
+  )
+
+  req.body.formsg.encryptedPayload = {
+    attachments: encryptedAttachments,
+    responses: filteredResponses.value as EncryptFormFieldResponse[],
+    encryptedContent,
+    version: req.body.version,
+  }
+
+  return next()
+}
+
+/**
+ * Moves encrypted payload present in req.body to req.body.formsg.encryptedPayload.
+ * Should only be used for the old storage mode submission endpoint (/api/v3/forms/:formId/submissions/encrypt).
+ */
+export const moveEncryptedPayload: EncryptSubmissionMiddlewareHandlerType =
+  async (req, res, next) => {
+    if (req.body.formsg.encryptedPayload) {
+      return res.send(new EncryptedPayloadExistsError())
+    }
```

Comment:
Is there any way we can enforce this in typing 🤔 Returning at runtime would be too late

Reply:
I don't see a way to enforce this through typing actually. Even for the `EncryptSubmissionMiddlewareHandlerType` used for `moveEncryptedPayload`, `encryptedPayload` is `undefined` before and defined after, but using the current typing does not stop stuff at compile time.

Reply:
thanks! i investigated this a bit more, indeed it does not seem possible with express middleware